File: Python/Data Structures/Segment Tree.py
# Under Testing
class Node():
  __slots__ = ['x', 'i', 'lazy']
  def __init__(self):
    self.x = 10**9+1; self.i = 0; self.lazy = 0
  def __repr__(self):
    return f"Node(x:{self.x}, i:{self.i}, lazy:{self.lazy})"

def copy(a):
  res = assign(a.x, a.i)
  return res

def merge(a, b):
  res = copy(a)
  if a.x > b.x: res = copy(b)
  return res

def assign(x, i):
  res = Node()
  res.x = x
  res.i = i
  return res

def apply(i, val):
  t[i].x += val
  if i<N: t[i].lazy += val

def build(a):
  global N, H
  n = len(a)
  N = 1; H = 0
  while N<n:  # N= 2^x & N>= n
    N *=2; H +=1
  t = [Node() for _ in range(2*N)]
  for i in range(n):
    t[N+i] = assign(a[i], i)
  for i in range(N-1, 0, -1):
    t[i] = merge(t[i<<1], t[i<<1 |1])
  return t

def push(t, p):
  for s in range(h, 0, -1):
    i = p >> s
    if t[i].lazy != 0:
      apply(i<<1, t[i].lazy)
      apply(i<<1|1, t[i].lazy)
      t[i].lazy = 0

def propagate(t, p):
  while p>1:
    p>>=1
    t[p].x = merge(t[p<<1], t[p<<1 |1]).x +t[p].lazy

# [l, r)
def update_rng(t, l, r, v):
  l +=N; r +=N
  l0,r0 = l,r
  push(t, l); push(t, r-1)  # NEW
  while l<r:
    if l&1==1:
      apply(l, v)
      l +=1
    if r&1==1:
      r-=1
      apply(r, v)
    l>>=1; r>>=1
  propagate(t, l0)
  propagate(t, r0-1)

def update(t, i, v):
  i +=N
  t[i] = assign(v, i-N)
  while i>1:
    i //=2
    t[i] = merge(t[i<<1], t[i<<1 |1])
# [l, r)
def query_rng(t, l, r):
  l +=N; r +=N
  push(t, l); push(t, r-1)  # NEW
  ansL = Node(); ansR = Node()
  while l<r:
    if l&1==1:
      ansL = merge(ansL, t[l])
      l +=1
    if r&1==1:
      r-=1
      ansR = merge(t[r], ansR)
    l>>=1; r>>=1
  return merge(ansL, ansR)
# [l, r]
def query_idx(t, l, r, k):
  L, R = 0, N-1
  s = [(1, L, R)]
  while s:
    i,L,R = s.pop()
    if t[i].x < k: continue
    if R<l or L>r: continue
    if L==R:
      return L
    mid = (L+R)//2
    if not(R<l or mid+1>r) and t[i<<1|1].x >= k:
      s.append((i<<1|1, mid+1, R))
    if not(mid<l or L>r) and t[i<<1].x >= k:
      s.append((i<<1, L, mid))
  return -1

File: C++/Data Structures/Segment Tree.cpp
using ll = long long;
const ll LAZY_DEF = 0;

struct node {
	ll x = 1e18;  ll idx=-1;  ll lazy=LAZY_DEF;  };

class Seg {
public:
	ll n;  vector<node> seg;
	Seg(ll n) {
		this->n = n;  seg.resize(4LL*(n +1LL));
	}
	// change here
	node combine(node &a, node &b) {
		node res = a;
		if (b.x < res.x) res = b;
		return res;
	}
	// change here
	node assign(ll x, ll idx) {
		node res;
		res.x = x;
		res.idx = idx;
		return res;
	}
	// change here
	void update_logic(ll idx, ll tl, ll tr, ll val) {
		seg[idx].x += val;
		seg[idx].lazy += val;
	}
	node no_overlap_return() {
		node res;
		return res;
	}
	void propagate(ll &idx, ll &tl, ll &tr) {
		if (seg[idx].lazy != LAZY_DEF and tl != tr) {
			ll mid = tl + ((tr - tl) >> 1LL);
			update_logic(2*idx +1, tl, mid, seg[idx].lazy);
			update_logic(2*idx +2, mid+1, tr, seg[idx].lazy);
			seg[idx].lazy = LAZY_DEF;
		}
	}
	void build_seg(ll idx, ll tl, ll tr, vector<ll> &a) {
		if (tl == tr) {
			seg[idx] = assign(a[tl], tl);
			return;
		}
		ll mid = tl + ((tr - tl) >> 1LL);
		build_seg(2*idx +1, tl, mid, a);
		build_seg(2*idx +2, mid+1, tr, a);
		seg[idx] = combine(seg[2*idx +1], seg[2*idx +2]);
	}
	node query_seg(ll idx, ll tl, ll tr, ll &l, ll &r) {
		propagate(idx, tl, tr);
		if (tl >= l and tr <= r) return seg[idx];
		if (tr < l or tl > r) return no_overlap_return();
		ll mid = tl + ((tr - tl) >> 1LL);
		node left = query_seg(2*idx +1, tl, mid, l, r);
		node right = query_seg(2*idx +2, mid+1, tr, l, r);
		return combine(left, right);
	}
	void update_seg(ll idx, ll tl, ll tr, ll &l, ll &r, ll &x) {
		propagate(idx, tl, tr);
		if (tl > r or tr < l) return;
		if (tl >= l and tr <= r) {
			update_logic(idx, tl, tr, x);
			return;
		}
		ll mid = tl + ((tr - tl) >> 1LL);
		update_seg(2*idx +1, tl, mid, l, r, x);
		update_seg(2*idx +2, mid+1, tr, l, r, x);
		seg[idx] = combine(seg[2*idx +1], seg[2*idx +2]);
	}
	int find_seg(ll idx, ll tl, ll tr, ll &l, ll &r, ll k) {
		if (tl > r or tr < l) return -1;
		if (seg[idx].x < k) return -1;
	    if (tl == tr) return tl;
	    propagate(idx, tl, tr);
	    ll mid = tl + ((tr - tl) >> 1LL);
	    ll res = find_seg(2*idx +1, tl, mid, l, r, k);
	    if (res != -1) return res;
	    return find_seg(2*idx +2, mid+1, tr, l, r, k - seg[2*idx +1].x);
	}

	void build(vector<ll> &a) {
		build_seg(0, 0, n-1, a);
	}
	node query(ll l, ll r) {
		return query_seg(0, 0, n-1, l, r);
	}
	void update(ll l, ll r, ll val) {
		update_seg(0, 0, n-1, l, r, val);
	}
	int find(ll l, ll r, ll k) {
    	return find_seg(0, 0, n-1, l, r, k);
  	}
};
